samples/shrubbery/reader.cc

```diff
@@ -0,0 +1,128 @@
+#include "shrubbery.h"
+#include "wf.h"
+
+namespace shrubbery
+{
+
+  auto err(const NodeRange& r, const std::string& msg)
+  {
+    return Error << (ErrorMsg ^ msg) << (ErrorAst << r);
+  }
+
+  auto err(Node node, const std::string& msg)
+  {
+    return Error << (ErrorMsg ^ msg) << (ErrorAst << node);
+  }
+
+  PassDef check_parsing()
+  {
+    return {
+      "check parsing",
+      wf_check_parser,
+      dir::topdown,
+      {
+        // An empty block followed by alternatives is ignored
+        ((T(Block) << End) * (T(Alt)[Alt])) >>
+          [](Match& _) { return _(Alt); },
+
+        (T(Block) << (T(Group)[Group] << T(Alt))) >>
+          [](Match& _) { return Seq << *_[Group]; },
+
+        // An empty group caused by a semicolon is ignored
+        (In(Semi) * ((T(Group) << End))) >>
+          [](Match&) { return Seq ^ ""; },
+
+        // Commas must separate (non-empty) groups
+        (T(Comma) << End)[Comma] >>
+          [](Match& _) { return err(_[Comma], "Comma does not separate groups"); },
+
+        (In(Comma) * (T(Group) << End)[Group]) >>
+          [](Match& _) { return err(_[Group], "Comma does not separate groups"); },
+
+        // A comma can only appear inside a paren, brace or bracket
+        ((--In(Paren, Brace, Bracket)) * T(Comma)[Comma]) >>
+          [](Match& _) { return err(_[Comma], "Commas can only separate groups in parentheses/braces/brackets"); },
+
+        // Opener-closer pairs must have comma-separated groups
+        (In(Paren, Brace, Bracket) * Any * Any)[Group] >>
+          [](Match& _) { return err(_[Group], "Groups in parentheses/braces/brackets must be comma separated"); },
+
+        // Opener-closer pairs cannot have semicolon-separated groups
+        (In(Paren, Brace, Bracket, Comma) * T(Semi))[Semi] >>
+          [](Match& _) { return err(_[Semi], "Semicolons cannot separate groups in parentheses/brackets/braces. Use commas."); },
+
+        // Blocks cannot be empty, except immediately under opener-closer pairs
+        // and as the only term in a top-level group
+        (--(In(Paren, Brace, Bracket, Comma, File))) * ((T(Group) << ((!T(Block))++ * (T(Block)[Block] << End)))) >>
+          [](Match& _) { return err(_[Block], "Blocks may not be empty"); },
+
+        In(File) * (T(Group) << (((!T(Block)) * (!T(Block))++ * (T(Block)[Block] << End) * End))) >>
+          [](Match& _) { return err(_[Block], "Blocks may not be empty"); },
+
+        // Alternatives cannot be empty
+        T(Alt)[Alt] << End >>
+          [](Match& _) { return err(_[Alt], "Alternatives may not be empty"); },
+      }
+    };
+  }
+
+  // Merge alternatives into a single Alt node containing blocks
+  PassDef merge_alternatives()
+  {
+    return {
+      "merge alternatives",
+      wf_alternatives,
+      dir::topdown,
```

Comment:
I think this could be turned into a `dir::once` pass by adding `++` to the captures below? I'll give an example of what I mean.